scene_graph/vrd_tutorial.py

```diff
@@ -0,0 +1,335 @@
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_carry_subject(x):
+    if x.object_category == "person":
+        if x.subject_category in ["chair", "bike", "snowboard", "motorcycle", "horse"]:
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_person(x):
+    if x.subject_category != "person":
+        return OTHER
+    return ABSTAIN
+
+
+# Distance-based LFs
+@labeling_function()
+def LF_ydist(x):
+    if x.subject_bbox[3] < x.object_bbox[3]:
+        return OTHER
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_dist(x):
+    if np.linalg.norm(np.array(x.subject_bbox) - np.array(x.object_bbox)) <= 1000:
+        return OTHER
+    return ABSTAIN
+
+
+# Size-based LF
+@labeling_function()
+def LF_area(x):
+    subject_area = (x.subject_bbox[1] - x.subject_bbox[0]) * (
+        x.subject_bbox[3] - x.subject_bbox[2]
+    )
+    object_area = (x.object_bbox[1] - x.object_bbox[0]) * (
+        x.object_bbox[3] - x.object_bbox[2]
+    )
+
+    if subject_area / object_area <= 0.5:
+        return OTHER
+    return ABSTAIN
+
+
+# %% [markdown]
+# Note that the labeling functions have varying empirical accuracies and coverages. Because of the imbalance in how the classes are defined, labeling functions that label the `OTHER` class have higher coverage than labeling functions for `RIDE` or `CARRY`. This reflects the distribution of classes in the dataset as well.
+
+# %%
+from snorkel.labeling.apply import PandasLFApplier
+lfs = [
+    LF_ride_object,
+    LF_ride_rare_object,
+    LF_carry_object,
+    LF_carry_subject,
+    LF_person,
+    LF_ydist,
+    LF_dist,
+    LF_area,
+]
+
+applier = PandasLFApplier(lfs)
+L_train = applier.apply(train_df)
+L_valid = applier.apply(valid_df)
+
+# %%
+from snorkel.labeling.analysis import LFAnalysis
+
+Y_valid = valid_df.label.values
+lf_names = [lf.name for lf in lfs]
+
+lf_analysis = LFAnalysis(L_valid)
+lf_analysis.lf_summary(Y_valid, lf_names=lf_names)
+
+# %% [markdown]
+# ## 3. Train Label Model
+# We now train a multi-class `LabelModel` to assign training labels to the unalabeled training set.
+
+# %%
+from snorkel.labeling.model import LabelModel
+
+label_model = LabelModel(cardinality=3, verbose=True)
+label_model.fit(L_train, seed=123, lr=0.01, log_freq=10, n_epochs=100)
+
+# %%
+label_model.score(L_valid, Y_valid, metrics=["f1_micro"])
+
+# %% [markdown]
+# ## 4. Train a Classifier
+# You can then use these training labels to train any standard discriminative model, such as [a state-of-the-art ResNet](https://github.com/KaimingHe/deep-residual-networks), which should learn to generalize beyond the LF's we've developed!
+
+# %% [markdown]
+# #### Create DataLoaders for Classifier
+
+# %%
+from snorkel.classification.data import DictDataLoader
+from scene_graph.model import FlatConcat, SceneGraphDataset, WordEmb, init_fc
+
+#change to "scene_graph/data/VRD/sg_dataset/sg_train_images" for full set
```

Comment:
suggestion to clean up: we can set this based on environment variable `os.environ["TRAVIS"]` in utils.py, and then `import TRAIN_DIR`